tsec-http4s/src/main/scala/tsec/authentication/JWKPublicKeyRSAAuthenticator.scala

```diff
@@ -0,0 +1,197 @@
+package tsec.authentication
+
+import java.security.KeyFactory
+import java.security.spec.RSAPublicKeySpec
+import java.time.Instant
+import java.util.concurrent.atomic.AtomicReference
+
+import cats.data.OptionT
+import cats.effect.{Effect, Sync}
+import cats.syntax.all._
+import fs2.Stream
+import fs2.async.Ref
+import io.circe.Decoder
+import io.circe.generic.auto._
+import io.circe.parser.decode
+import org.http4s.circe._
+import org.http4s.client.blaze._
+import org.http4s.headers.Authorization
+import org.http4s.{AuthScheme, Credentials, EntityDecoder, Request, Response, Uri}
+import tsec.common.{SecureRandomId, _}
+import tsec.jws.signature.{JWSSigCV, JWTSig}
+import tsec.signature.jca.{JCASigTag, SigPublicKey}
+
+import scala.concurrent.duration.FiniteDuration
+
+final case class AugmentedJWK[A, I](
+    id: SecureRandomId,
+    jwt: JWTSig[A],
+    identity: I,
+    expiry: Instant,
+    lastTouched: Option[Instant]
+)
+
+final case class Modulus(value: BigInt) extends AnyVal
+final case class Exponent(value: BigInt) extends AnyVal
+final case class JWK(kid: String, kty: String, use: String, n: Modulus, e: Exponent)
+
+final case class JWKS(keys: List[JWK])
+
+class KeyRegistry[F[_], A: JCASigTag](uri: Uri, minFetchDelay: FiniteDuration)(implicit E: Effect[F]) {
+
+  implicit val jwksDecoder: EntityDecoder[F, JWKS] = jsonOf[F, JWKS]
+  implicit val modulusDecoder: Decoder[Modulus]    = Decoder.decodeString.map(s => Modulus(BigInt(1, s.base64UrlBytes)))
+  implicit val exponentDecoder: Decoder[Exponent]  = Decoder.decodeString.map(s => Exponent(BigInt(1, s.base64UrlBytes)))
+
+  private val keys      = new AtomicReference(Map[String, SigPublicKey[A]]())
+  private val lastFetch = new AtomicReference(none[Instant])
+
+  def getPublicKey(id: String): F[Option[SigPublicKey[A]]] = {
+    getKey(id).map {
+      case None if shouldFetch() =>
+        for {
+          jwks      <- fetchJwks(uri)
+          k         <- E.delay(jwks.keys.map(jwk => (jwk.kid, convert(jwk))).toMap)
+          _         <- E.delay(keys.set(k))
+          _         <- E.delay(lastFetch.set(Instant.now().some))
+          key       <- E.delay(k.get(id))
+        } yield key
+      case _ => E.delay(keys.get().get(id))
+    }.flatten
+  }
+
+  private def fetchJwks(uri:Uri) = {
+    val s = for {
+      client <- Http1Client.stream[F]()
```

Comment:
This creates an Http1Client _every single time_. 

Make the class take a `Client[F]`, otherwise this will be _excruciatingly slow_, as it will allocate a connection pool and then close it after you make your request. It makes no sense to allocate and then shutdown a connection pool for a single request every time.